### PR TITLE
Nqf0059: Diabetics Hemoglobin (9%)

### PR DESCRIPTION
**Pull request by deepson-maitri**
## Describe your changes
This PR aims to have the Diabetics Quality Measure to be in the Tuva repo. This has been achieved by updating the seeds, creating new denominator, numerator and exclusion models, leading upto the summary tables.
Additionally a macro has been created to filter columns on the basis of provided regex.

## How has this been tested?
Patients meeting the criteria have been selected and lab results of such patients have been manually updated for the labs table. Following is a snippet of test performed.
image

## Reviewer focus
Denominator and numerator data models

## Checklist before requesting a review
- [x] I have updated the version number in dbt_project.yml file to reflect the release number of this PR
- [x] I have updated the docs files (by running dbt docs generate/serve and copying the necessary files into the docs folder)
- [x] I have commented my code as necessary
- [x] I have added at least one Github label to this PR
- [x] My code follows style guidelines
- [ ] (Optional) I have recorded a Loom to explain this PR

### CODE REVIEW

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+), visits_encounters as (
+    select patient_id
+         , coalesce(encounter.encounter_start_date,encounter.encounter_end_date) as min_date
+         , coalesce(encounter.encounter_end_date,encounter.encounter_start_date) as max_date
+    from {{ref('quality_measures__stg_core__encounter')}} encounter
+    inner join {{ref('quality_measures__int_nqf0059__performance_period')}} as pp
+        on coalesce(encounter.encounter_end_date,encounter.encounter_start_date) >= pp.performance_period_begin
+        and  coalesce(encounter.encounter_start_date,encounter.encounter_end_date) <= pp.performance_period_end
+    where lower(encounter_type) in (
```
Comment (sarah-tuva) [inline models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql:29]: Can you add 'telehealth' to this list? That's the equivalent of 'telephone visits' from the measure's value sets.

**Review by sarahmcmorgan — CHANGES_REQUESTED** [review]
Hey Deepson, great work on this! I left a few comments that need to be addressed. 

I also noticed that none of the Diabetes models were added to the [quality_measures_models.yml](https://github.com/deepson-maitri/tuva/blob/main/models/quality_measures/quality_measures_models.yml) which means they are getting created under the user's default schema. We want these created in the quality_measures schema. This also has performance issues since they are getting created as views and not tables. Adding them to the YML with the correct config options will fix these issues.

**Comment by sarah-tuva** [comment]
I'm seeing an issue with the numerator model when running it on real patient data. The lab_result core table includes non-numeric values. We'll need to filter these non-numeric values out before applying the result filters. See the screenshot below of the error.

image

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        , performance_period_begin
+        , performance_period_end
+        , measure_id
+        , measure_name
+        , measure_version
+        , evidence_date
+        , 1 as performance_flag
+        , 1 as numerator_flag
+    from valid_patients
+    where result > 9.0
```
Comment (sarah-tuva) [inline models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql:133]: BigQuery is complaining about the data type of "result". Adding this type cast to all of the readings CTEs resolves it.
```
 where cast(result as {{ dbt.type_numeric() }}) > 9.0
```

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+, diabetics_codes as (
+
+    select
+          code
+        , code_system
+    from {{ ref('quality_measures__value_sets') }}
+    where lower(concept_name) in (
+        'diabetes',
+        'hba1c laboratory test',
+        'nutrition services'
```
[the remark this answers is not in the document]
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql]: `nutrition services` has been moved to `visit_codes` CTE

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    from qualifying_labs
+    qualify row_number() over(partition by patient_id order by evidence_date desc) = 1
+
+)
+
+, qualifying_patients as (
+
+    select
+        denominator.*
+        , recent_readings.* exclude patient_id
```
[the remark this answers is not in the document]
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql]: `exclude` has been removed and required columns has been listed from the `recent_reading` cte

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+    select
+          patient_id
+        , performance_period_begin
+        , performance_period_end
+        , measure_id
+        , measure_name
+        , measure_version
+        , evidence_date
+        , 1 as performance_flag
```
[the remark this answers is not in the document]
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql]: `performance_flag` has been removed from the numerator data model.

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        , cast(measure_name as {{ dbt.type_string() }}) as measure_name
+        , cast(measure_version as {{ dbt.type_string() }}) as measure_version
+        , cast(evidence_date as date) as evidence_date
+        , cast(numerator_flag as integer) as numerator_flag
+        , cast(performance_flag as integer) as performance_flag
+    from numerator
+
+)
+
+select * from add_data_types
```
[the remark this answers is not in the document]
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql]: All the columns have been listed.

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        , measure_version
+        , evidence_date
+        , 1 as performance_flag
+        , 1 as numerator_flag
+    from valid_patients
+    where result > 9.0
+
+)
+
+, readings_less_than_7 as (
```
[the remark this answers is not in the document]
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql]: removed redundant logic

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql (hunk 1 above)
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql:30]: `telehealth` has been added

**Comment by deepson-maitri** [comment]
> I'm seeing an issue with the numerator model when running it on real patient data. The lab_result core table includes non-numeric values. We'll need to filter these non-numeric values out before applying the result filters. See the screenshot below of the error.
> 
> image

such values will be filtered out by the applied regex.

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+{%- set performance_period_end -%}
+(
+
+    select 
+        performance_period_end
+    from {{ ref('quality_measures__int_nqf0059__performance_period') }}
+
+)
+{%- endset -%}
+
```
[the remark this answers is not in the document]
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql]: This logic has been replaced by using `denominator` model

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    from labs
+    inner join hba1c_test_code
+      on ( labs.normalized_code = hba1c_test_code.code
+       and labs.normalized_code_type = hba1c_test_code.code_system )
+      or ( labs.source_code = hba1c_test_code.code
+       and labs.source_code_type = hba1c_test_code.code_system )
+    left join denominator
+        on labs.patient_id = denominator.patient_id
+    where coalesce(collection_date,result_date) <= denominator.performance_period_end
+        and regexp_like(labs.result, '[+-]?([0-9]*[.])?[0-9]+')
```
Comment (sarah-tuva) [inline models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_numerator.sql:60]: Nice solution for filtering non-numeric labs. 👏

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+
+)
+
+, qualifying_patients as (
+
+    select
+        distinct
+          diabetic_conditions.patient_id
+        , patients_with_age.* exclude patient_id
```
Comment (sarah-tuva) [inline models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql:152]: Sorry, I missed this in my first review. Can you change this to list out the columns from patients_with_age?

**Comment by sarah-tuva** [comment]
> > I'm seeing an issue with the numerator model when running it on real patient data. The lab_result core table includes non-numeric values. We'll need to filter these non-numeric values out before applying the result filters. See the screenshot below of the error.
> > image
> 
> such values will be filtered out by the applied regex. `[+-]?([0-9]*[.])?[0-9]+`

I was just testing in BigQuery and found it doesn't have the regexp_like function. I'll keep testing to find another option for this.

File: models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql (hunk 10 above)
Comment (deepson-maitri) [reply models/quality_measures/intermediate/nqf0059_diabetes_hemoglobin_a1c/quality_measures__int_nqf0059_denominator.sql]: exclude clause has been removed

**Comment by deepson-maitri** [comment]
> > > I'm seeing an issue with the numerator model when running it on real patient data. The lab_result core table includes non-numeric values. We'll need to filter these non-numeric values out before applying the result filters. See the screenshot below of the error.
> > > image
> > 
> > 
> > such values will be filtered out by the applied regex. `[+-]?([0-9]*[.])?[0-9]+`
> 
> I was just testing in BigQuery and found it doesn't have the regexp_like function. I'll keep testing to find another option for this.

A macro has been added to this PR to support other databases. The macro's directory is `macros/apply_regex.sql`

File: macros/apply_regex.sql [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+{%- macro snowflake__apply_regex(column_name, regex) -%}
+
+    regexp_like({{ column_name }}, '{{ regex }}')
+
+{%- endmacro -%}
+
+{%- macro bigquery__apply_regex(column_name, regex) -%}
+
+    regexp_contains(column_name, r'{{ regex }}')
```
Comment (sarah-tuva) [inline macros/apply_regex.sql:21]: This column_name reference is missing brackets. Getting an error when testing in BigQuery.

image


```suggestion
    regexp_contains({{ column_name }}, r'{{ regex }}')
```

**Review by sarahmcmorgan — APPROVED** [review]

**Pull request merged**